# Projet de modélisation d'un mix énergétique issu 100% d'ENR

## IV. Simulations du modèle complet sur l'optimisation du coût du mix énergétique

### Imports

In [5]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import csv
import time
import sys

# Imports pour pyomo
import pyomo.environ as pyo
from pyomo.opt import SolverFactory

from ipywidgets import interactive, interact_manual
import ipywidgets as widgets

from matplotlib.figure import Figure
from matplotlib.backends.backend_agg import FigureCanvas

### 1. Utils

In [7]:
def affichage(simu):
    print(" --- Q ---")
    for i in simu.model.Q:
        print(i," : ",pyo.value(simu.model.Q[i]))   
    print(" --- S ---")   
    for i in simu.model.S:
        print(i," : ",pyo.value(simu.model.S[i]))
    print(" --- volume ---")
    for i in simu.model.volume:
        print(i," : ",pyo.value(simu.model.volume[i]))
    print(" --- cost --- ")
    print(simu.cost())

### 2. Simulations du modèle complet
Optimisation des variables Q, S et VOLUME  
Optimisation du mix énergétique

### <span style="color:blue">-------------------------------------------------------------------------------------------------------------------------------------------------------</span>
### <span style="color:blue">A exécuter si on n'a pas le dossier Simulations_completes</span>

In [2]:
from complete_model import complete_model

##### RTE Demand

In [6]:
simu = complete_model()
simu.load_param("inputs/", "vre_profiles2006.csv", "demand2050_rte.csv")
simu.run()
affichage(simu)

Chargement des paramètres ...
vre_profiles2006.csv
demand2050_rte.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  120.0
pv  :  153.59165
river  :  7.5
lake  :  12.855
biogas  :  37.009384
phs  :  6.3936009
battery  :  33.875077
methanation  :  10.828085
 --- S ---
phs  :  6.2
battery  :  33.875077
methanation  :  10.991424
 --- volume ---
phs  :  135.5
battery  :  186.77956
methanation  :  14356.455
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             29.214223                           0.000041


In [8]:
res = simu.write_results('simu_vre2006_rte', "vre_profiles2006.csv", "demand2050_rte.csv")

Ecriture des résultats ...


##### Négawatt Demand

In [9]:
simu0 = complete_model()
simu0.load_param("inputs/", "vre_profiles2006.csv", "demand2050_negawatt.csv")
simu0.run()
affichage(simu0)

Chargement des paramètres ...
vre_profiles2006.csv
demand2050_negawatt.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  9.2239321
onshore  :  35.15871
pv  :  69.845503
river  :  7.5
lake  :  12.855
biogas  :  16.977882
phs  :  3.8188521
battery  :  7.6920888
methanation  :  6.8141513e-06
 --- S ---
phs  :  6.2
battery  :  7.6920888
methanation  :  -1.1730846e-06
 --- volume ---
phs  :  135.5
battery  :  35.682393
methanation  :  0.0011505466
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             10.495222                           0.000034


In [10]:
res0 = simu0.write_results('simu_vre2006_nega', "vre_profiles2006.csv", "demand2050_negawatt.csv")

Ecriture des résultats ...


##### ADEME Demand

In [11]:
simu1 = complete_model()
simu1.load_param("inputs/", "vre_profiles2006.csv", "demand2050_ademe.csv")
simu1.run()
affichage(simu1)

Chargement des paramètres ...
vre_profiles2006.csv
demand2050_ademe.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  70.230296
pv  :  113.03217
river  :  7.5
lake  :  12.855
biogas  :  36.785692
phs  :  5.7047562
battery  :  13.343201
methanation  :  10.745931
 --- S ---
phs  :  6.2
battery  :  13.343201
methanation  :  7.5578362
 --- volume ---
phs  :  135.5
battery  :  64.603839
methanation  :  8181.5733
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             19.801001                           0.000038


In [12]:
res1 = simu1.write_results('simu_vre2006_ademe', "vre_profiles2006.csv", "demand2050_ademe.csv")

Ecriture des résultats ...


##### SImulation demand 2006 with production year 2006

In [13]:
simu2 = complete_model()
simu2.load_param("inputs/", "vre_profiles2006.csv", "demand2006.csv")
simu2.run()
affichage(simu2)

Chargement des paramètres ...
vre_profiles2006.csv
demand2006.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  96.928942
pv  :  111.63747
river  :  7.5
lake  :  12.855
biogas  :  38.02888
phs  :  5.8387458
battery  :  16.849963
methanation  :  11.192881
 --- S ---
phs  :  6.2
battery  :  16.849963
methanation  :  8.8512386
 --- volume ---
phs  :  135.5
battery  :  89.873494
methanation  :  12928.6
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             23.276694                           0.000039


In [14]:
res2 = simu2.write_results('simu_vre2006_d2006', "vre_profiles2006.csv", "demand2006.csv")

Ecriture des résultats ...


##### SImulation new production year with typical days

In [15]:
simu3 = complete_model()
simu3.load_param("inputs/", "vrenew_typical_days_new_trend_2006.csv", "demand2050_rte.csv")
simu3.run()
affichage(simu3)

Chargement des paramètres ...
vrenew_typical_days_new_trend_2006.csv
demand2050_rte.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  120.0
pv  :  133.3451
river  :  7.5
lake  :  12.855
biogas  :  12.562493
phs  :  6.9183834
battery  :  27.34422
methanation  :  23.683933
 --- S ---
phs  :  6.2
battery  :  27.34422
methanation  :  10.154642
 --- volume ---
phs  :  135.5
battery  :  139.28016
methanation  :  25525.281
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0              27.68108                            0.00004


In [17]:
res3 = simu3.write_results('simu_td2006_rte', "vrenew_typical_days_new_trend_2006.csv", "demand2050_rte.csv")

Ecriture des résultats ...


In [18]:
simu4 = complete_model()
simu4.load_param("inputs/", "vrenew_typical_days_new_trend_2006.csv", "demand2050_ademe.csv")
simu4.run()
affichage(simu4)

Chargement des paramètres ...
vrenew_typical_days_new_trend_2006.csv
demand2050_ademe.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  68.86535
pv  :  102.8958
river  :  7.5
lake  :  12.855
biogas  :  100.0
phs  :  6.0456196
battery  :  11.597403
methanation  :  100.0
 --- S ---
phs  :  6.2
battery  :  11.597403
methanation  :  6.1049549
 --- volume ---
phs  :  135.5
battery  :  52.117602
methanation  :  9344.6234
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             18.879165                           0.000038


In [19]:
res4 = simu4.write_results('simu_td2006_ademe', "vrenew_typical_days_new_trend_2006.csv", "demand2050_ademe.csv")

Ecriture des résultats ...


In [20]:
simu5 = complete_model()
simu5.load_param("inputs/", "vrenew_typical_days_new_trend_2006.csv", "demand2050_negawatt.csv")
simu5.run()
affichage(simu5)

Chargement des paramètres ...
vrenew_typical_days_new_trend_2006.csv
demand2050_negawatt.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  8.3844081
onshore  :  39.8228
pv  :  56.005471
river  :  7.5
lake  :  12.855
biogas  :  13.13401
phs  :  4.4422619
battery  :  4.3007106
methanation  :  -4.0408764e-08
 --- S ---
phs  :  6.2
battery  :  4.3007106
methanation  :  -8.8696267e-07
 --- volume ---
phs  :  135.5
battery  :  13.206684
methanation  :  0.00043261572
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             10.029627                           0.000034


In [21]:
res5 = simu5.write_results('simu_td2006_nega', "vrenew_typical_days_new_trend_2006.csv", "demand2050_negawatt.csv")

Ecriture des résultats ...


In [22]:
simu6 = complete_model()
simu6.load_param("inputs/", "vrenew_typical_days_new_trend_2006.csv", "demand2006.csv")
simu6.run()
affichage(simu6)

Chargement des paramètres ...
vrenew_typical_days_new_trend_2006.csv
demand2006.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  94.531141
pv  :  99.077313
river  :  7.5
lake  :  12.855
biogas  :  12.374523
phs  :  6.4891041
battery  :  13.863512
methanation  :  22.722664
 --- S ---
phs  :  6.2
battery  :  13.863512
methanation  :  8.8184852
 --- volume ---
phs  :  135.5
battery  :  63.80723
methanation  :  23079.58
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             22.197027                           0.000039


In [23]:
res6 = simu6.write_results('simu_td2006_d2006', "vrenew_typical_days_new_trend_2006.csv", "demand2006.csv")

Ecriture des résultats ...


In [19]:
simu6 = complete_model()
simu6.load_param("inputs/", "vrenew_fourier_profile.csv", "demand2050_ademe.csv")
simu6.run()
affichage(simu6)

Chargement des paramètres ...
vrenew_fourier_profile.csv
demand2050_ademe.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  74.405287
pv  :  85.995392
river  :  7.5
lake  :  12.855
biogas  :  18.903713
phs  :  5.9335914
battery  :  7.2473175
methanation  :  24.198772
 --- S ---
phs  :  6.2
battery  :  7.2473175
methanation  :  4.5323527
 --- volume ---
phs  :  135.5
battery  :  25.050954
methanation  :  4452.0403
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             18.220577                           0.000037


In [20]:
res6 = simu6.write_results('simu_fourier_ademe', "vrenew_fourier_profile.csv", "demand2050_ademe.csv")

Ecriture des résultats ...


In [21]:
simu6 = complete_model()
simu6.load_param("inputs/", "vrenew_fourier_profile.csv", "demand2050_rte.csv")
simu6.run()
affichage(simu6)

Chargement des paramètres ...
vrenew_fourier_profile.csv
demand2050_rte.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  119.99604
pv  :  128.05585
river  :  7.5
lake  :  12.855
biogas  :  100.0
phs  :  5.8281103
battery  :  24.957722
methanation  :  100.0
 --- S ---
phs  :  6.2
battery  :  24.957722
methanation  :  10.987952
 --- volume ---
phs  :  135.5
battery  :  129.78697
methanation  :  7967.1378
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             27.412874                            0.00004


In [22]:
res6 = simu6.write_results('simu_fourier_rte', "vrenew_fourier_profile.csv", "demand2050_rte.csv")

Ecriture des résultats ...


In [23]:
simu6 = complete_model()
simu6.load_param("inputs/", "vrenew_fourier_profile.csv", "demand2050_negawatt.csv")
simu6.run()
affichage(simu6)

Chargement des paramètres ...
vrenew_fourier_profile.csv
demand2050_negawatt.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  16.123495
onshore  :  28.772462
pv  :  52.677768
river  :  7.5
lake  :  12.855
biogas  :  17.757525
phs  :  4.9345641
battery  :  2.2152677
methanation  :  0.0
 --- S ---
phs  :  6.2
battery  :  2.2152677
methanation  :  -2.174031e-08
 --- volume ---
phs  :  135.5
battery  :  6.092101
methanation  :  3.2791935e-05
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             10.043325                           0.000034


In [24]:
res6 = simu6.write_results('simu_fourier_nega', "vrenew_fourier_profile.csv", "demand2050_negawatt.csv")

Ecriture des résultats ...


In [25]:
simu6 = complete_model()
simu6.load_param("inputs/", "vrenew_fourier_profile.csv", "demand2006.csv")
simu6.run()
affichage(simu6)

Chargement des paramètres ...
vrenew_fourier_profile.csv
demand2006.csv
Initialisation des sets...
Définition des variables à optimiser ...
Ajout des contraintes ...
Optimisation ...
Simulation du modèle faite avec succès ! 
 --- Q ---
offshore  :  20.0
onshore  :  98.540984
pv  :  82.813888
river  :  7.5
lake  :  12.855
biogas  :  15.052778
phs  :  5.9824917
battery  :  9.6037475
methanation  :  40.854729
 --- S ---
phs  :  6.2
battery  :  9.6037475
methanation  :  6.1540465
 --- volume ---
phs  :  135.5
battery  :  41.27661
methanation  :  4734.1399
 --- cost --- 
   COST (billion euros)  Cost per MWh produced (euros/MWh)
0             21.268006                           0.000038


In [27]:
res6 = simu6.write_results('simu_fourier_d2006', "vrenew_fourier_profile.csv", "demand2006.csv")

Ecriture des résultats ...


FileExistsError: [Errno 17] File exists: 'simu_fourier_d2006/simu_fourier_d2006_init_data.txt'

#### Résultats 

vre_profiles2006.csv  
Demande 2006  
COST (billion euros) 23.276694

vrenew_fourier.csv  
Demande 2006  
COST (billion euros) 24.529635

vrenew_moyennes.csv  
Demande 2006  
COST (billion euros) 22.999705

vrenew_moyennes_solaire.csv  
Demande 2006  
COST (billion euros) 23.298378  

vrenew_typical_days.csv  
Demande 2006  
COST (billion euros) 21.747864   

// avec modélisation des tendances 20.760234 avec 18 clusters       
// avec modélisation des tendances 20.373473 avec 12 clusters     21.206023   
// avec modélisation des tendances 19.313955 avec 6  clusters

### <span style="color:blue">-------------------------------------------------------------------------------------------------------------------------------------------------------</span>

### 3. Affichage des résultats avec le modèle complet

In [4]:
from complete_model import show_simu

In [5]:
noms_simu = ['2006_ademe', '2006_rte', '2006_nega',
             '2010_ademe', '2010_rte', '2010_nega',
             '2011_ademe', '2011_rte', '2011_nega',
             '2012_ademe', '2012_rte', '2012_nega',
             '2013_ademe', '2013_rte', '2013_nega',
             '2015_ademe', '2015_rte', '2015_nega',
             '2016_ademe', '2016_rte', '2016_nega',
             '2017_ademe', '2017_rte', '2017_nega',]
# Dashboard
interactive_plot = interact_manual(show_simu, week=widgets.IntText(value = 1), name = noms_simu, opti=False)
interactive_plot

interactive(children=(IntText(value=1, description='week'), Dropdown(description='name', options=('2006_ademe'…

<function complete_model.show_simu(week, name, opti=True)>

In [6]:
noms_simu = ["td2006_rte", "td2006_ademe", "td2006_nega"]
# Dashboard
interactive_plot = interact_manual(show_simu, week=widgets.IntText(value = 1), name = noms_simu, opti=False)
interactive_plot

interactive(children=(IntText(value=1, description='week'), Dropdown(description='name', options=('td2006_rte'…

<function complete_model.show_simu(week, name, opti=True)>

#### Selon les scénarios de demande : 
Demande RTE > Demande ADEME > Demande NégaWatt 

#### Selon les semaines : 
En hiver, on a beaucoup d'éolien onshore, on arrive à subvenir à la demande en journée lorsqu'il y a du soleil pour alimenter les panneaux solaires. Lorsqu'il fait nuit on alimente le réseau avec la méthanation, le biogas et les barages.  
Au printemps, on a plus de soleil et un peu moins de vent donc on stocke l'énergie solaire en surplus la journée dans les batteries pour alimenter le réseau la nuit. Lorsqu'il y a peu de vent, on utilise l'énergie des barages et des rivières. A l'inverse, on a parfois beaucoup de vent qu'on stocke par méthanation.  
En été, c'est le solaire qui alimente fortement le réseau le jour ; la nuit le réseau est alimenté par batteries, les barages ainsi que le rivières.  
En automne, on a une configuration similaire à celle en hiver.